drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -755,11 +755,72 @@ static int siwx91x_get_version(const struct device *dev, struct wifi_version *pa
 	return 0;
 }
 
+static int siwx91x_get_connected_ap_beacon_interval(void)
+{
+	sl_wifi_operational_statistics_t sl_stat = { };
+	sl_status_t status;
+
+	status = sl_wifi_get_operational_statistics(SL_WIFI_CLIENT_INTERFACE, &sl_stat);
+	if (status != SL_STATUS_OK) {
+		return status;
+	}
+
+	return sys_get_le16(sl_stat.beacon_interval);
+}
+
+static int siwx91x_convert_li_bcn_unit_to_ms(uint16_t bcn_li)
+{
+	uint16_t beacon_interval;
+	uint16_t ms_li;
+
+	beacon_interval = siwx91x_get_connected_ap_beacon_interval();
+	if (beacon_interval < 0) {
+		LOG_ERR("Failed to get beacon interval: 0x%x", beacon_interval);
+		return -EAGAIN;
+	}
+
+	ms_li = bcn_li * beacon_interval;
+	if (ms_li > SIWX91X_PS_MAX_LISTEN_INTERVAL_MS) {
+		LOG_WRN("Listen interval exceeded, capping to %d",
+			SIWX91X_PS_MAX_LISTEN_INTERVAL_MS);
+		ms_li = SIWX91X_PS_MAX_LISTEN_INTERVAL_MS;
+	}
+
+	return ms_li;
+}
+
+static int siwx91x_schedule_listen_interval_wakeup(const struct device *dev, uint16_t bcn_li)
+{
+	struct siwx91x_dev *sidev = dev->data;
+	sl_wifi_performance_profile_t *sl_ps_profile = &sidev->sl_ps_profile;
+	uint16_t ms_li;
+
+	if (sidev->state != WIFI_STATE_COMPLETED) {
+		LOG_WRN("Device not connected, listen interval reflect after connection");
+		return -EINVAL;
+	}
```

Comment:
Take care of your design. I assume the job of this function is to fill `sl_ps_profile`. The state of the device is checked by the callers.

In addition, this function does not not "schedule" anything. It just fills `sl_ps_profile` with proper values.

Probably you want:
```
int siwx91x_set_listen_interval(sl_wifi_performance_profile_t *sl_ps_profile, uint16_t bcn_li);
```